tensorflow_datasets/core/features/rle_feature/rle/tf_impl.py

```diff
@@ -0,0 +1,101 @@
+"""Tensorflow encode/decode/utility implementations."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+
+major, minor = (int(v) for v in tf.__version__.split('.')[:2])
+is_over_1_12 = major >= 1 and minor > 12
+
+
+def ensure_supported_version(error_identifier):
+  if not is_over_1_12:
+    raise NotImplementedError(
+      "%s not implemented for version < 1.13" % error_identifier)
+
+
+def brle_length(brle):
+  return tf.reduce_sum(brle)
+
+
+def rle_length(rle):
+  return tf.reduce_sum(rle[1::2])
+
+
+def brle_logical_not(brle):
+  """Get the BRLE encoding of the `logical_not`ed decoding of brle.
+
+  Equivalent to `encode(tf.logical_not(decode(brle)))` but highly optimized.
+  Actual implementation just pads brle with a 0 on each end.
+
+  Args:
+    brle: rank 1 int tensor of
+  """
+  return tf.pad(brle, [[1, 1]])
+
+
+def brle_to_dense(brle, vals=None):
+  """Decode binary run length encoded data.
+
+  Args:
+    brle: binary run length encoded data, 1D tensor of int dtype
+    vals (optional): length-2 tensor/array/tuple made up of
+        (false_val, true_val).
+
+  Returns:
+    Dense representation, rank-1 tensor with dtype the same as vals or `tf.bool`
+      if `vals is None`.
+  """
+  ensure_supported_version('`brle_to_dense`')
+  brle = tf.convert_to_tensor(brle)
+  if not brle.dtype.is_integer:
+    raise TypeError("`brle` must be of integer type.")
+  if not brle.shape.ndims == 1:
+    raise TypeError("`brle` must be a rank 1 tensor.")
+  enc_false, enc_true = tf.unstack(tf.reshape(brle, (-1, 2)), axis=1)
+
+  n_false = tf.reduce_sum(enc_false)
+  n_true = tf.reduce_sum(enc_true)
+  if vals is None:
+    false_vals = tf.zeros((n_false,), dtype=tf.bool)
+    true_vals = tf.ones((n_true,), dtype=tf.bool)
+  else:
+    vals = tf.convert_to_tensor(vals)
+    if vals.shape.ndims != 1:
+      raise ValueError("`vals` must be rank 1 tensor.")
+    if vals.shape[0] != 2:
+      raise ValueError("`vals` must have exactly 2 entries.")
+
+    false_val, true_val = tf.unstack(vals, axis=0)
+    false_vals = tf.fill([n_false], false_val)
+    true_vals = tf.fill([n_true], true_val)
+
+  ragged_false = tf.RaggedTensor.from_row_lengths(false_vals, enc_false)
+  ragged_true = tf.RaggedTensor.from_row_lengths(true_vals, enc_true)
+
+  ragged_all = tf.concat([ragged_false, ragged_true], axis=1)
+  return ragged_all.values
+
+
+def tf_repeat(values, counts):
```

Comment:
This is possible to do in TF and we, for now, are not allowing `py_function` usage in the `tf.data` pipeline. This ensures that the input pipelines are high-performance and work everywhere.

Reply:
I suspected this might be the case (but given I had all the other pieces for an `RleFeature` I felt it was worth putting in with disclaimers in the README).

Can you direct me to a working implementation of `tf.repeat`? [This issue](https://github.com/tensorflow/tensorflow/issues/8246) is still open, and while there was a pull request with a custom kernel implementation it was abandoned (the reviewer seemed to think it could be implemented using existing ops... but I never worked out how). The implementation given in the issue doesn't seem to work when there are counts of 0 - this is unlikely to be an issue for `int64` encodings, but is certainly an issue with, say, `uint8` encodings (which aren't included here, so we might be able to get away with it... but I'd feel more comfortable making it more general).

Reply:
So here's something that may work for this instance though doesn't capture the full generality of `np.repeat`. Does this work for you?

https://gist.github.com/rsepassi/d2367b4355828575b57fe86c28c3f1cf

Reply:
... well that's pretty obvious now you mention it. I've put something similar in. Would love to make it a `foldl` rather than a `while_loop`, but that's because I just don't feel comfortable with seeing `repeats[i]`. Not sure if [this](https://github.com/tensorflow/tensorflow/pull/26030) will get through but figured I'd give it a shot anyway.

A similar thing works with BRLE decoding, so I've put a check in there to switch between a `loop`y implementation or `RaggedTensor` version depending on `RaggedTensor` availability. Any thoughts on speed? That also removes any `tf_comp` issues.